### DISCUSSION TITLE
Has anyone deployed oxygen to a docker container ?

### DISCUSSION THREAD

**Discussion by IoannisMaras** [OxygenFramework/Oxygen.jl#103] [General]
I manage to deploy it and it runs fine , but no matter what I cannot send an http request to oxygen . So I want wondering if anyone managed to do it and if so how?

**Comment by ndortega** [comment]
@IoannisMaras 
Below is a **very** simple dockefile you can use to deploy your application. This was something I've used to quickly test deployments with docker.

Fair warning this image is not optimized and pulls in all your packages along with Julia itself. So this image ends up being around 500mb on my computer. I'd recommend looking into [StaticCompiler.jl](https://github.com/tshort/StaticCompiler.jl) and the [PackageCompiler.j](https://github.com/JuliaLang/PackageCompiler.jl) to create smaller executables which should result in a much smaller image.

```Dockerfile
# Base image
FROM julia:1.6

# Set the working directory
WORKDIR /app

# Copy the Project.toml and Manifest.toml files into the image to manage dependencies
COPY Project.toml Manifest.toml ./

# Install the dependencies in a separate layer to improve caching
RUN julia --project -e 'using Pkg; Pkg.instantiate()'

# Copy the source code into the image
COPY src/ ./src

# Expose the desired port
EXPOSE 8080

# Set the entrypoint to start the application
ENTRYPOINT ["julia", "--project", "src/main.jl"]
```

Commands to build & run
```bash
docker build -t oxygen-app .      
docker run -p 8080:8080 oxygen-app
```

My project structure

![image](https://user-images.githubusercontent.com/22534471/234593051-148ff563-01a3-4836-abe7-04b4e09b961d.png)

**Reply by IoannisMaras** [reply]
I am new to using docker but something is not working for me. I will leave all details below and maybe someone can me

First of all this is my dockerfile and main.jl

![image](https://user-images.githubusercontent.com/34172738/234606415-f812c817-50a6-47ee-bf44-6726061bd58d.png)
![image](https://user-images.githubusercontent.com/34172738/234606505-20f77447-7902-4f1c-a009-5dc2b8a40891.png)

When I run docker everything seems fine as you can see in the image below :
![image](https://user-images.githubusercontent.com/34172738/234606644-073f6fda-9a05-44c5-8680-ba901fe9a0c1.png)

But no matter what I try to hit (0.0.0.0:8050 , localhost:8050 , ip of container:8050) Nothing works and I just get 500

**Reply by ndortega** [reply]
what command are you using to run the container? Are the ports connected?

`docker run -p 8080:8080 oxygen-app`

In this example, I'm connecting my local 8080 port to the 8080 port inside the container.

**Reply by IoannisMaras** [reply]
I used docker run -p 8080:8050 juliaapp and as you can see in the background it does not work . Maybe i should get rid of some files check file by file it something is causing a problem . But if i run it localhost it works just fine

![image](https://user-images.githubusercontent.com/34172738/234646715-ecd82e39-e8c1-4c49-bde6-f5e1453aa071.png)

**Reply by ndortega** [reply]
Just to clarify, if you're getting a 500 from the server then that's at least a good sign (did you get the standard error message?). It means the connection between your localhost and the container is working fine. 

If you are getting the 500 server message from oxygen:
- check to see how you're loading your files into your web server. 
- Try commenting out all other routes and adding a simple '/' request handler to make sure everything runs fine

If you are not getting the 500 from the oxygen server:
- then I recommend creating a small test project (like the one I shared) to test the docker integration.

**Reply by IoannisMaras** [reply]
Managed to make it work , seems like the way you went with .toml files is the right one

just sharing my julia 1.8.5 if anyone else in the future

Thank you very much for your time

```Dockerfile
#Base image
FROM julia:1.8.5

#Set the working directory
WORKDIR /JuliaServer

#Copy the Project.toml and Manifest.toml files into the image to manage dependencies
COPY Project.toml Manifest.toml ./

#Enable multithreading for Julia
ENV JULIA_NUM_THREADS=4

#Install the dependencies in a separate layer to improve caching
RUN julia --project -e 'using Pkg; Pkg.instantiate()'

#Copy the source code into the image
COPY src/ ./src

#Expose the desired port
EXPOSE 8050

#Set the entrypoint to start the application
ENTRYPOINT ["julia", "--project=.", "src/JuliaServer.jl"]
```

**Comment by arlowhite** [comment]
Here's our Dockerfile https://github.com/open-AIMS/ReefGuideAPI.jl/blob/main/Dockerfile
A bit more complicated and takes longer to build, but compiles to different CPU targets so the image doesn't need to precompile in different environments.
`ENV JULIA_CPU_TARGET=x86_64;haswell;skylake;skylake-avx512;tigerlake`
